pennylane/devices/qubit/simulate.py

```diff
@@ -484,65 +473,50 @@ def gather_non_mcm(circuit_measurement, measurement, samples):
+            f"Native mid-circuit measurement mode does not support {type(circuit_measurement)} measurements."
         )
     return new_meas
 
 
-@singledispatch
-def gather_mcm(measurement, mv, samples, counter):  # pylint: disable=unused-argument
+def gather_mcm(measurement, samples):
     """Combines, gathers and normalizes several measurements with non-trivial measurement values.
 
     Args:
         measurement (MeasurementProcess): measurement
-        mv (MeasurementValue): measurement value
         samples (List[dict]): Mid-circuit measurement samples
-        counter (Counter): Measurement value counts
 
     Returns:
         TensorLike: The combined measurement outcome
     """
-    raise ValueError(
-        f"Native mid-circuit measurement mode does not support {measurement.__class__.__name__} measurements."
-    )
-
-
-@gather_mcm.register
-def _(measurement: CountsMP, mv, samples, counter):  # pylint: disable=unused-argument
-    new_samples = {}
-    if len(samples) - counter[mv.measurements[0]] > 0:
-        new_samples[0] = len(samples) - counter[mv.measurements[0]]
-    if counter[mv.measurements[0]] > 0:
-        new_samples[1] = counter[mv.measurements[0]]
-    return new_samples
-
-
-@gather_mcm.register
-def _(measurement: ExpectationMP, mv, samples, counter):  # pylint: disable=unused-argument
-    return np.mean(np.array([dct[mv.measurements[0]] for dct in samples]))
-
-
-@gather_mcm.register
-def _(measurement: ProbabilityMP, mv, samples, counter):  # pylint: disable=unused-argument
-    p1 = counter[mv.measurements[0]] / float(len(samples))
-    return np.array([1 - p1, p1])
-
-
-@gather_mcm.register
-def _(measurement: SampleMP, mv, samples, counter):  # pylint: disable=unused-argument
-    return np.array([dct[mv.measurements[0]] for dct in samples])
-
-
-@gather_mcm.register
-def _(measurement: VarianceMP, mv, samples, counter):  # pylint: disable=unused-argument
-    return qml.math.var(np.array([dct[mv.measurements[0]] for dct in samples]))
+    mv = measurement.mv
+    if isinstance(measurement, (CountsMP, ProbabilityMP, SampleMP)) and isinstance(mv, Sequence):
+        wires = qml.wires.Wires(range(len(mv)))
+        mcm_samples = list(
+            np.array([m.concretize(dct) for dct in samples]).reshape((-1, 1)) for m in mv
+        )
+        mcm_samples = np.concatenate(mcm_samples, axis=1)
+        meas_tmp = measurement.__class__(wires=wires)
+        return meas_tmp.process_samples(mcm_samples, wire_order=wires)
+    mcm_samples = np.array([mv.concretize(dct) for dct in samples]).reshape((-1, 1))
+    use_as_is = len(mv.measurements) == 1
+    if use_as_is:
+        wires, meas_tmp = mv.wires, measurement
+    else:
+        if isinstance(measurement, (ExpectationMP, VarianceMP)):
+            mcm_samples = mcm_samples.ravel()
+        wires = qml.wires.Wires(0)
+        meas_tmp = measurement.__class__(wires=wires)
```

Comment:
can you explain what's happening here? I'm a little worried when we arbitrarily use a wire like `0`, and I am unsure of what happens if you make something like `ExpectationMP(wires=Wires(0))`. It technically creates an object, but it doesn't feel like it should be valid.

Reply:
When we have a composite measurement, say `qml.count((op=(m0 - 2 * m1) * m2 + 7))` where `m0`, `m1`, `m2` are MCMs on wires 0, 1 and 0 respectively, `mv.wires` will be `<Wires = [0, 1]>`. But `mcm_samples = np.array([mv.concretize(dct) for dct in samples]).reshape((-1, 1))` is not a 2-wire sample, it is like a 1-wire sample sampling from the interval `[5, 8]`. I thus trick the code by introducing `wires = qml.wires.Wires(0)` so I can use `process_samples` down the line.

Reply:
hmm.. this is definitely tricking the `SampleMeasurement` interface into doing something it isn't supposed to do. It works, but it makes me nervous. This seems like the interface isn't meeting our needs, so I won't block the PR on it but I wanna make sure people know what's happening.

Reply:
Let me add the explanation above in a comment then. I could also try to hijack `process_samples` to use the observable branch of `_samples_to_counts` instead of the `wires` branch. Would that be better, or just more of the same?

Reply:
imo more of the same. a comment would be perfect for now 😄